### DISCUSSION TITLE
How to do server side form validation with nextjs

### DISCUSSION THREAD

**Discussion by ricjohn-acosta** [vercel/next.js#15359] [Help] [answered]
Hey everyone,

I'm currently building a CRUD app and I opted for the built-in server that nextjs provides instead of integrating express and I'm wondering how would I make sure that the data that my front-end is sending to my endpoints are valid? Right now I'm thinking of doing the validation on the api page, is this a wise thing to do?

**Comment by jamesmosier** [comment, marked as answer]
Hi @ricjohn-acosta. Yes, doing your validation in your API route would be a great option. Additionally you could use a library like [yup](https://github.com/jquense/yup) for both client side and server side validations.

**Reply by oddjones** [reply]
I for one would love to see an example of this - does anybody know of one?

**Reply by MOTORIST** [reply]
```
// helper
export async function validation<T = Record<string, any>>(
  scheme: SchemaOf<T>,
  data: Record<string, any> | null
) {
  try {
    await scheme.validate(data, { abortEarly: false });
    return { isValid: true, errors: null };
  } catch (error) {
    const { errors } = error;
    return { isValid: false, errors };
  }
}

interface UserModel {
  username: string;
  email: string;
  password: string;
}

// common (client - server)
export const userValidationSchema: yup.SchemaOf<UserModel> = yup.object().shape({
  username: yup.string().required(),
  email: yup.string().email().required(),
  password: yup.string().required(),
});

async function handle(req, res) {
  const { method, body } = req;
  const { errors } = await validation(userValidationSchema, body);

  if (errors) {
      res.status(422).json({ errors });
      return;
   }
}


```
Client example:  [formik](https://formik.org/docs/guides/validation#validationschema)

**Comment by giovannipds** [comment]
You'll probably need a middleware. See [API Middlewares / Connect/Express middleware support](https://nextjs.org/docs/api-routes/api-middlewares#connectexpress-middleware-support) for a clear understanding. You could also consider `yup` as suggested by @jamesmosier together.
